Do not generate a silent assertion for an address in a transition state.

shane-tomlinson opened the pull request:
@lloyd, @ozten, @seanmonstar - need discussion and review.

Commit:
Do not generate a silent assertion for an address in a transition state. Update backend to correctly mark addresses as "secondary" upon verification.

* When calling /wsapi/complete_reset.js or /wsapi/complete_transition.js, immediately flip the type of the email address to "secondary"
* Add explanation of silent assertions.

fixes #3515

jaredhirsch commented:
@shane-tomlinson holy :hankey: dude, that was quick! Glad I was on the right track, if going 20x slower than your analysis :-D

Commit:
Add Selenium tests to check for transition states.

Commit:
Add backend tests for transition_no_password and transition_to_secondary cases to ensure email type is set to secondary and state changed to known.

shane-tomlinson commented:
From the email I sent yesterday, emails converted from a primary-secondary were only marked as secondary when /wsapi/cert_key was generated. This meant if a user closed the original dialog before they verified their email address, when they re-opened the dialog the address would _still_ be in a transition state.

To avoid this, I made a change to the backend - as soon as a user verifies ownership of an email address with the fallback IdP, mark the address to a secondary. The user verified. They signaled their desire to convert.

I had to update one backend test, in address-info-test.js, line 373. The test checks the return status of /wsapi/address_info and is labelled "returns 'known' and not 'offline'". @ozten, you made the change on line 373 from "known" to "transition_to_secondary", I changed it back to "known".

I believe our behavior has been incorrect and could have led to confusion if users tried to convert their account from primary->secondary and closed the original dialog before verifying their ownership of the address. If users did this on an RP that used the .get API, the user would have continually been locked in the "transition_to_secondary" state because /wsapi/cert_key would never have been called.

What, if any, repercussions does my change have?

shane-tomlinson commented:
@ozten, @lloyd, @seanmonstar - can I get review/discussion on this?

Commit:
Add a transition-to-secondary-forgot-password Selenium test.

This is for a user who has to transition to a secondary but forgot their password.

ozten commented on lib/wsapi/complete_user_creation.js:
```diff
@@ -95,7 +95,16 @@ exports.process = function(req, res) {
+            // Unless the type is changed to "secondary", /wsapi/address_info
+            // will return "transition_to_secondary" as the address state.
+            // The user has verified the address. The transition is complete.
+            // Get rid of the transition state.
```
This comment is long and confusing. Can we make it more concise? Write it to your future self.

ozten commented on resources/static/common/js/user.js:
```diff
@@ -1178,17 +1188,12 @@ BrowserID.User = (function() {
      */
     checkForInvalidCerts: function(email, info, done) {
       function clearCert(email, idInfo) {
+        delete idInfo.priv;
```
Nice.

ozten commented on resources/static/common/js/user.js:
```diff
@@ -1464,24 +1476,86 @@ BrowserID.User = (function() {
+          return complete(onComplete, null, null);
+
+        User.resetCaches();
+        User.addressInfo(loggedInEmail, function(info) {
```
This is expensive. I'm guessing you've tried hard to avoid this. Down the road we could 
- add an expiration to our cache, 
- Do a series of addressInfo calls while the dialog is idle
- ???
  or something to optimize this, or some other optimization to avoid the extra network hit.

ozten commented on tests/address-info-test.js:
```diff
@@ -370,7 +370,7 @@ suite.addBatch({
           var r = JSON.parse(r.body);
           assert.equal(r.type, "secondary");
-          assert.equal(r.state, "transition_to_secondary");
+          assert.equal(r.state, "known");
```
Sorry about that. I don't have much of this swapped into my head, but this looks good.
